pkg/antctl/raw/multicluster/join.go

```diff
@@ -0,0 +1,282 @@
+
+	return command
+}
+
+func joinRunE(cmd *cobra.Command, args []string) error {
+	var err error
+	if err = joinOpts.validateAndComplete(); err != nil {
+		return err
+	}
+	k8sClient, err := common.NewClient(cmd)
+
+	memberClusterNamespace := joinOpts.MemberClusterOpt.Namespace
+	memberClusterID := joinOpts.MemberClusterOpt.ClusterID
+	memberClusterSet := joinOpts.MemberClusterOpt.ClusterSet
+	createdRes := []*unstructured.Unstructured{}
+	defer func() {
+		if err != nil {
+			fmt.Fprintf(cmd.OutOrStdout(), "Failed to join the ClusterSet, begin to delete the created resources.\n")
+			if err := common.Rollback(cmd, k8sClient, createdRes); err != nil {
+				fmt.Fprintf(cmd.OutOrStdout(), "Failed to rollback: %v\n", err)
+			}
+		}
+	}()
+	err = common.CreateClusterClaim(cmd, k8sClient, memberClusterNamespace, memberClusterSet, memberClusterID, &createdRes)
+	if err != nil {
+		return err
+	}
+	err = common.CreateClusterSet(cmd, k8sClient, memberClusterNamespace, memberClusterSet, joinOpts.LeaderClusterOpt.Server, joinOpts.LeaderClusterOpt.Secret,
+		joinOpts.MemberClusterOpt.ClusterID, joinOpts.LeaderClusterOpt.ClusterID, joinOpts.LeaderClusterOpt.Namespace, &createdRes)
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(cmd.OutOrStdout(), "Waiting for member cluster ready\n")
+	if err = waitForMemberClusterReady(cmd, k8sClient); err != nil {
+		fmt.Fprintf(cmd.OutOrStdout(), "Failed to wait for member cluster ready: %v\n", err)
+		return err
+	}
+	fmt.Fprintf(cmd.OutOrStdout(), "Member cluster joined successfully\n")
+
+	return nil
+}
+
+func waitForMemberClusterReady(cmd *cobra.Command, k8sClient client.Client) error {
+	fmt.Fprintf(cmd.OutOrStdout(), "Waiting for ClusterSet \"%s\" in Namespace %s to be registered in leader cluster\n",
+		joinOpts.MemberClusterOpt.ClusterSet, joinOpts.MemberClusterOpt.Namespace)
+	secret := &v1.Secret{}
+	if err := k8sClient.Get(context.TODO(), client.ObjectKey{Name: joinOpts.LeaderClusterOpt.Secret, Namespace: joinOpts.MemberClusterOpt.Namespace}, secret); err != nil {
+		fmt.Fprintf(cmd.OutOrStdout(), "Failed to get secret \"%s\" in Namespace %s: %v\n", joinOpts.LeaderClusterOpt.Secret, joinOpts.MemberClusterOpt.Namespace, err)
+		return err
+	}
+
+	crtData, token, err := commonarea.GetSecretCACrtAndToken(secret)
+	if err != nil {
+		return err
+	}
+	config, err := clientcmd.BuildConfigFromFlags(joinOpts.LeaderClusterOpt.Server, "")
+	if err != nil {
+		return err
+	}
+	config.BearerToken = string(token)
+	config.CAData = crtData
+
+	remoteClient, err := client.New(config, client.Options{Scheme: multiclusterscheme.Scheme})
+	if err != nil {
+		return err
+	}
+
+	memberClusterAnnounce := "member-announce-from-" + joinOpts.MemberClusterOpt.ClusterID
+	fmt.Fprintf(cmd.OutOrStdout(), "Waiting for member cluster \"%s\" in Namespace %s in leader cluster\n", joinOpts.MemberClusterOpt.ClusterID, joinOpts.MemberClusterOpt.Namespace)
```

Comment:
Could we remove this one. I feel it duplicates the earlier one at line 202.

Reply:
Removed.